libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,719 @@
+    /// if a file is missing, consider showing the user the contents of the directory where the file was expected to be
+    /// and the full path of the file.
+    ///
+    /// If you are confident about what action needs to be taken to fix the error, you should include that in the error message.
+    /// Do not write a generic suggestion like "try again later" unless you are certain that the error is transient.
+    ///
+    /// If you detect something problematic but not bad enough to halt buildpack execution, consider using a [`BuildpackOutput::warning`] instead.
+    pub fn error(mut self, s: impl AsRef<str>) {
+        self.write_paragraph(&ANSI::Red, s);
+    }
+
+    /// Emit a warning message to the end user.
+    ///
+    /// A warning should be used to emit a message to the end user about a potential problem.
+    ///
+    /// Multiple warnings can be emitted in sequence. The buildpack author should take care not to overwhelm the end user with
+    /// unnecessary warnings.
+    ///
+    /// When emitting a warning, describe the problem to the user, if possible, and tell them how to fix it or where to look next.
+    ///
+    /// Warnings should often come with some disabling mechanism, if possible. If the user can turn off the warning,
+    /// that information should be included in the warning message. If you're confident that the user should not be able to
+    /// turn off a warning; use an [`BuildpackOutput::error`] instead
+    ///
+    /// Warnings will be output in a multi-line paragraph style. A warning can be emitted from any state except for [`state::NotStarted`].
+    #[must_use]
+    pub fn warning(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(&ANSI::Yellow, s);
+        self
+    }
+
+    /// Emit an important message to the end user.
+    ///
+    /// When something significant happens but is not inherently negative, you can use an important message. For example,
+    /// if a buildpack detects that the operating system or architecture has changed since the last build, it might not be a problem,
+    ///, but if something goes wrong, the user should know about it.
+    ///
+    /// Important messages should be used sparingly and only for things the user should be aware of but not necessarily act on.
+    /// If the message is actionable, consider using a [`BuildpackOutput::warning`] instead.
+    #[must_use]
+    pub fn important(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(&ANSI::BoldCyan, s);
+        self
+    }
+
+    fn write_paragraph(&mut self, color: &ANSI, s: impl AsRef<str>) {
+        let io = self.state.write_mut();
+
+        if !io.was_paragraph {
+            writeln_now(io, "");
+        }
+
+        writeln_now(
+            io,
+            ansi_escape::wrap_ansi_escape_each_line(
+                color,
+                prefix_lines(s.as_ref(), |_, line| {
+                    if line.chars().all(char::is_whitespace) {
```

Comment:
I see this change to `is_empty()` was reverted in 8be7f9069f71c1ff9505ed7b3f3d64569201cc0cto fix an additional case identified in a new test.

However, switching back to `line.chars().all(char::is_whitespace)` reintroduces a different bug.

I've fixed that bug by switching to `line.is_empty() || line == "\n"`, and updated the tests to catch this additional case, in:
5e20c4154d31fe55891419f1c479996af177c90c

(Feel free to mark this as resolved when you've read it; leaving it open for now otherwise it's hard to find/see my reply)

Reply:
Short: 👍🏻 I'm good with your change.

Long:
I reviewed the commit. I think this is something (hopefully) no one will ever care about or need to care about. Since it's whitespace either way, the end user should never see it. The primary driver of the original logic was test consistency which, btw, I like your testing strategy with indoc here. It's clear and robust.